onnxruntime/core/framework/session_state.h

```diff
@@ -217,9 +243,17 @@ class SessionState {
   std::vector<BufferUniquePtr>& GetMutableWeightsBuffers() { return weights_buffers_; }
   const NodeIndexInfo& GetNodeIndexInfo() const;
 
+  void UpdateToBeExecutedNodes(const std::vector<int>& fetch_mlvalue_idxs);
+  const std::unordered_set<NodeIndex>* GetToBeExecutedNodes(const std::vector<int>& fetch_mlvalue_idxs) const;
+
  private:
   ORT_DISALLOW_COPY_ASSIGNMENT_AND_MOVE(SessionState);
 
+  Status GeneratePatternGroupCache(
+      const std::vector<std::reference_wrapper<const TensorShape>>& input_shape,
```

Comment:
Is the purpose of this just to avoid the first execution not using the memory pattern planner block? If so, what's the cost or issue if it does?

Reply:
@souptc can you comment on this?

---
In reply to: [397032118](https://github.com/microsoft/onnxruntime/pull/3174#discussion_r397032118) [](ancestors = 397032118)

Reply:
yes, the idea is to optimize memory usage for training, we need to static planning all the memory reusing in the graph (due to dynamic shape, allocation planner didn't do a good job), that result in allocate 1 big chunk for whole iteration. If first iteration does not use memory pattern, it will result in more memory fraction in arena. You can image on a 16GB GPU, without memory pattern, the arena may result in a list of chunks with size from 1GB to 6GB. Then we can't create chunk with bigger size.

Reply:
ok. another option would be to just make the initial chunk size in BFCArena much larger than 1MB. 

---
In reply to: [400615777](https://github.com/microsoft/onnxruntime/pull/3174#discussion_r400615777) [](ancestors = 400615777)

Reply:
I didn't see how it solve the issue? are you saying adjust the initial chunk to size like 8GB? 
i don't feel that is good idea.

Reply:
If you do a large enough allocation upfront the smaller allocations from the first run without the memory pattern planner block can be re-used for the memory pattern planner block in the later runs (they would be coalesced back into the large chunk at the end of the run). 

Calculating the block size upfront is more accurate if you have shape information for all the intermediate outputs. Wondering what happens if you don't (e.g. IIRC Reshape nodes with dynamic input will result in missing shape info from the node down)? Do you allocate an initial block size based on the info you do have, run the model, and update the memory pattern planner block size at the end of the run now that you know exact sizes? If so, doesn't that still lead to a scenario where the first block is too small and can't be re-used?

There are pros and cons to either approach.

Reply:
From Scott's - I want to make the initial chunk size and max ‘fragmentation’ values configurable, but first we need to add a mechanism to allow that sort of low level configuration to be done without adding separate parameters to SessionOptions for every single one of them.

I’d say leave it for now as what is best can be very model and/or use-case dependent. 


---
In reply to: [405819374](https://github.com/microsoft/onnxruntime/pull/3174#discussion_r405819374) [](ancestors = 405819374)